common/types/overflow.go

```diff
@@ -222,57 +259,99 @@ func subtractTimeChecked(x, y time.Time) (time.Duration, bool) {
 
 	// Subtract seconds first, detecting any overflow.
-	sec, ok := subtractInt64Checked(sec1, sec2)
-	if !ok {
-		return time.Duration(0), false
+	sec, err := subtractInt64Checked(sec1, sec2)
+	if err != nil {
+		return time.Duration(0), err
 	}
 
 	// Nanoseconds cannot overflow as time.Time normalizes them to [0, 999999999].
 	nsec := nsec1 - nsec2
 
-	// We need to normalize nanoseconds to be positive and carry extra nanoseconds to seconds.
-	// Adapted from time.Unix(int64, int64).
-	if nsec < 0 || nsec >= int64(time.Second) {
-		// Add an extra second.
-		sec, ok = addInt64Checked(sec, nsec/int64(time.Second))
-		if !ok {
-			return time.Duration(0), false
-		}
-		nsec -= (nsec / int64(time.Second)) * int64(time.Second)
-		if nsec < 0 {
-			// Subtract an extra second
-			sec, ok = addInt64Checked(sec, -1)
-			if !ok {
-				return time.Duration(0), false
-			}
-			nsec += int64(time.Second)
-		}
-	}
-
 	// Scale seconds to nanoseconds detecting overflow.
-	tsec, ok := multiplyInt64Checked(sec, int64(time.Second))
-	if !ok {
-		return time.Duration(0), false
+	tsec, err := multiplyInt64Checked(sec, int64(time.Second))
+	if err != nil {
+		return time.Duration(0), err
 	}
 
 	// Lastly we need to add the two nanoseconds together.
-	val, ok := addInt64Checked(tsec, nsec)
-	if !ok {
-		return time.Duration(0), false
+	val, err := addInt64Checked(tsec, nsec)
+	if err != nil {
+		return time.Duration(0), err
 	}
 
-	return time.Duration(val), true
+	return time.Duration(val), nil
 }
 
 // subtractTimeDurationChecked performs subtraction with overflow detection of a time.Time and
-// time.Duration, returning the result of the subtraction if no overflow occurred as the first
-// return value and a bool indicating whether no overflow occurred as the second return value.
-func subtractTimeDurationChecked(x time.Time, y time.Duration) (time.Time, bool) {
+// time.Duration.
+//
+// If the operation fails due to overflow the error return value will be non-nil.
+func subtractTimeDurationChecked(x time.Time, y time.Duration) (time.Time, error) {
 	// The easiest way to implement this is to negate y and add them.
 	// x - y = x + -y
-	val, ok := negateDurationChecked(y)
-	if !ok {
-		return time.Time{}, false
+	val, err := negateDurationChecked(y)
+	if err != nil {
+		return time.Time{}, err
 	}
 	return addTimeDurationChecked(x, val)
 }
+
+// doubleToInt64Checked converts a double to an int64 value.
+//
+// If the conversion fails due to overflow the error return value will be non-nil.
+func doubleToInt64Checked(v float64) (int64, error) {
+	if math.IsInf(v, 0) || v <= float64(math.MinInt64) || v >= float64(math.MaxInt64) {
+		return 0, errIntOverflow
+	}
+	return int64(v), nil
+}
+
+// doubleToInt64Checked converts a double to a uint64 value.
+//
+// If the conversion fails due to overflow the error return value will be non-nil.
+func doubleToUint64Checked(v float64) (uint64, error) {
+	if math.IsInf(v, 0) || v < 0 || v >= doubleTwoTo64 {
```

Comment:
Couldn't you just use math.MaxUint64? math.Ldexp seems less readable.

Reply:
I've kept the same call pattern as is used in C++ for consistency (it uses `std::ldexp`)